### PR TITLE
Metabot with embeddings and caching

### PR DESCRIPTION
**Pull request by markbastian**
Large databases have been problematic for Metabot v2 since input token usage was ridiculously large. This PR uses embeddings to encode inputs and select the one or best input segment for model selection and native sql inference prompts.

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30166)

### CODE REVIEW

File: src/metabase/metabot.clj [hunk 1]
```diff
@@ -2,12 +2,13 @@
   "The core metabot namespace. Consists primarily of functions named infer-X,
   where X is the thing we want to extract from the bot response."
   (:require
-   [cheshire.core :as json]
-   [metabase.lib.native :as lib-native]
-   [metabase.metabot.client :as metabot-client]
-   [metabase.metabot.settings :as metabot-settings]
-   [metabase.metabot.util :as metabot-util]
-   [metabase.util.log :as log]))
+    [cheshire.core :as json]
```
Comment (dpsutton) [inline src/metabase/metabot.clj:5]: this indentation change seems weird. not sure which is right offhand but i don't like that the level changed. is this related to your config loss?

Comment (dpsutton) [reply src/metabase/metabot.clj:5]: From other namespaces, it should be underneath the `:`. (and getting this wrong creates painful merge conflicts that have no substance to them, just an extra space)

File: src/metabase/metabot.clj [hunk 2]
```diff
@@ -21,10 +22,10 @@
       ;; More interesting SQL merits a more interesting display
       (let [{:keys [prompt_template version] :as prompt} (metabot-util/create-prompt context)]
         {:template                (metabot-util/find-result
-                                   (fn [message]
-                                     (metabot-util/response->viz
-                                      (json/parse-string message keyword)))
-                                   (metabot-client/invoke-metabot prompt))
+                                    (fn [message]
```
Comment (dpsutton) [inline src/metabase/metabot.clj:25]: again some indentation changes with nothing of substance changed.

File: src/metabase/metabot/client.clj [hunk 3]
```diff
@@ -1,15 +1,16 @@
...
+    [cheshire.core :as json]
+    [clojure.core.memoize :as memoize]
+    [metabase.metabot.settings :as metabot-settings]
+    [metabase.util.log :as log]
+    [wkok.openai-clojure.api :as openai.api]))
 
 (set! *warn-on-reflection* true)
 
-(def ^:dynamic *bot-endpoint*
+(def ^:dynamic *create-chat-completion-endpoint*
```
Comment (dpsutton) [inline src/metabase/metabot/client.clj:11]: usually better to do something like

```clojure
(defn- default-chat-completion-endoint [params options] ...)
(def ^:dynamic *create-chat-completion-endpoint* default-chat-completion-endpoint)
```

bonus points for putting arglist metadata on the dynamic var.

Comment (markbastian) [reply src/metabase/metabot/client.clj:11]: Done

File: src/metabase/metabot.clj (hunk 1 above)
Comment (dpsutton) [reply src/metabase/metabot.clj:5]: I don't believe the indentation is fixed. The diff here still shows all of the rows indented a single space. And comparing to other namespaces it is more indented

```clojure
(ns metabase.api.card-test
  "Tests for /api/card endpoints."
  (:require
   [cheshire.core :as json]
   [clojure.set :as set]
```

File: src/metabase/metabot/client.clj [hunk 4]
```diff
@@ -1,26 +1,32 @@
 (ns metabase.metabot.client
   (:require
-   [cheshire.core :as json]
-   [metabase.metabot.settings :as metabot-settings]
-   [metabase.util.log :as log]
-   [wkok.openai-clojure.api :as openai.api]))
+    [cheshire.core :as json]
```
Comment (dpsutton) [inline src/metabase/metabot/client.clj:3]: again indentation

File: src/metabase/metabot/client.clj [hunk 5]
```diff
@@ -60,3 +66,37 @@
...
+          {:keys [prompt_tokens]} :usage} (*create-embedding-endpoint*
+                                            {:model model
+                                             :input prompt}
+                                            {:api-key      (metabot-settings/openai-api-key)
+                                             :organization (metabot-settings/openai-organization)})]
+     {:prompt    prompt
+      :embedding embedding
+      :tokens    prompt_tokens}))
+  ([prompt]
+   (create-embedding "text-embedding-ada-002" prompt)))
```
Comment (dpsutton) [inline src/metabase/metabot/client.clj:102]: not super necessary but i think we like to extract things like `"text-embedding-ada-002"` into a top level def for clarity.

File: test/metabase/metabot_test.clj [hunk 6]
```diff
@@ -94,10 +108,59 @@
...
+              (partial=
+               denormalized-model
+               (metabot/match-best-model context))))))))))
+
+(deftest infer-native-sql-test
+  (testing "Test the 'plumbing' of the infer-native-sql function. The actual invocation of the remote bot is dynamically rebound."
+    (mt/with-temporary-setting-values [is-metabot-enabled true]
+      (mt/dataset sample-dataset
+        (let [user_prompt "Show me all of my reviews data"
+              db          (t2/select-one Database :id (mt/id))
```
Comment (dpsutton) [inline test/metabase/metabot_test.clj:133]: you can just use `(mt/db)` for this

**Review by dpsutton — APPROVED** [review]
a few notes and the indentation is still not correct. Contrast with your changes to `metabase.metabot-test`
image

File: src/metabase/metabot/client.clj (hunk 5 above)
Comment (markbastian) [reply src/metabase/metabot/client.clj]: Done

File: src/metabase/metabot.clj (hunk 1 above)
Comment (markbastian) [reply src/metabase/metabot.clj]: Good catch. I thought I'd looked at every file.
